File: Vizualizations/PieVisualization/1_prepareData.py
# This scripts prepares the calculated classification results
# from GPC and CNN for pie and ranking visualization.
#
# In this script, we load the trainign matrix from the GPC
# and CNN, get the lengthscale values from the GPC and convert
# them. The convertion is done be invert the values, store the 
# inverted lengthscale values (=relevance) and get the mean values 
# over all k's in the k-fold cross validation.
# 
# We store two relevance values, namely the mean of all entries in the
# relevance matrix and the mean of overall col mean. 
#
# This code is available under a GPL v3.0 license and comes without
# any explicit or implicit warranty.
#
# (C) Wilfried Woeber 2020 <wilfried.woeber@technikum-wien.at>
import numpy as np 
import os
import pandas as pd
#------------------------#
#--- global functions ---#
#------------------------#
# Create a folder for a given dataset
def mkdir(classifier):
    os.system("mkdir data/data_"+classifier) #Creates a folder for given classifier
# Returns the name of the classifiers data folder
def getdir(classifier):
    return("data/data_"+classifier)
# Converts data to a pandas DF with names
def convertARD(M):
    myDF = pd.DataFrame(relevance_local)    #Convert to pandas DF with false col name
    #--- Create column names ---#
    colnames = ['ard_1']    #Default name
    for i in range(2,M.shape[1]+1):
        colnames.extend(['ard_'+str(i)])    #Extend name list
    myDF.columns = colnames     #Changed colum names
    return(myDF)
#------------------------------#
#--- Main processing starts ---#
#------------------------------#
nr_classes = 6          #Number of populations
k_fold = range(0,10)    #Range of ks
iterations = 10         #Number of iterations
os.system("mkdir data")	#Created folder for generated data
#--- Create initial data ---#
data_1 = {'name': "BGPLVM_noResampling_manual", 'folder':"../../GPC/GPLVM/",        'isCNN': False, 'nrFeatures':14}
data_2 = {'name': "BGPLVM_noResampling_top14",  'folder':"../../GPC/GPLVM_full/",   'isCNN': False, 'nrFeatures':14}
data_3 = {'name': "Procrustes_noResampling",    'folder':"../../GPC/Procrustes/",   'isCNN': False, 'nrFeatures':28}
data_4 = {'name': "CNN_Aug_noResampling",       'folder':"../../CNN/Classification/Augmented/", 'isCNN': True, 'nrFeatures':0}
DF = pd.DataFrame(data=data_1,index=[0])
DF = DF.append(data_2,ignore_index=True)
DF = DF.append(data_3,ignore_index=True)
DF = DF.append(data_4,ignore_index=True)
#--- Prepare system ---#
os.system("rm -rf data/data*")  #Remove old data if exists
#---------------------------------#
#--- Prepare data and get data ---#
#---------------------------------#
for i in range(0,len(DF)): #Loop over all defined classifiers
    data = DF.loc[i,]   #Get classifier info
    name = data[2]      #Name of classifier
    path = data[0]      #Path to classifier data
    isCNN= bool(data[1])    #See if classifier was an CNN
    nrFeatures=data[3]  #Number of features of classifier
    #--- Do processing ---#
    # 1: Create data folder
    mkdir(name)     #Create folder with classifier name
    # 2: Calculate relevance
    RELEVANCE_SUM = np.zeros((iterations,nrFeatures))   #Create memory for all values
    for ITERATION in range(0,iterations):       #We got 10 iterations -> 0..9
        DF_classifier_result = pd.read_csv(path+str(ITERATION)+'_dataMatrix.csv')
        #--- Store here ---#
        DF_classifier_result.to_csv(getdir(name)+"/"+str(ITERATION)+"_classifierResult.csv", index=False)
        #--- Get relevance if available (no CNN) ---#
        if(not isCNN):
            relevance_local = np.zeros((len(k_fold),nrFeatures))            #Memory for later use - the col mean values
            relevance_per_population = np.zeros((nr_classes,nrFeatures))    #Memory for mean of each entry in the relevance matrix
            #To get the relevance we need to loop over all k's 
            for k in k_fold:    #Loop over all k's
                LS_of_k = np.genfromtxt(path+str(ITERATION)+'_foldNr_'+str(k)+'_LS.csv',delimiter=' ')
                #LS_of_k consists of the LS values of the GPC --> mxd matrix, where m is the number of populations ans d is the data dimension (here, 14)
                inv_LS_of_k = 1./LS_of_k            #Invert the lengthscale values
                colMeans = np.mean(inv_LS_of_k,0)   #Get col mean
                #--- Store data ---#
                relevance_per_population = relevance_per_population+inv_LS_of_k #Add relevance values together
                relevance_local[k,:] = colMeans                                 #Store col means
            #--- Post process relevance values ---#
            MEAN_relevance_per_population = relevance_per_population/len(k_fold)    #Get mean relevance values
            np.savetxt(getdir(name)+"/"+str(ITERATION)+"_MeanRelevance_perPopulation.csv",MEAN_relevance_per_population, delimiter=',')
            DF_ARD = convertARD(relevance_local)                        #Convert to pandas DF
            DF_ARD_mean = pd.DataFrame((np.mean(DF_ARD,0))).transpose() #Get col means
            DF_ARD_mean.to_csv(getdir(name)+"/"+str(ITERATION)+"_Relevance.csv",index=False)
            RELEVANCE_SUM[ITERATION,:]=np.array(DF_ARD_mean)
        #End if is CNN
    #End Iteration loop
    if(not isCNN):
        np.savetxt(getdir(name)+"/_RelevanceSUM.csv",RELEVANCE_SUM, delimiter=',')

File: Vizualizations/PieVisualization/1_prepareData.R
#This file converts the GPC results to PS format
rm(list = ls())    #Remove old variables
library(readr)
#------------------------#
#--- Global functions ---#
#------------------------#
mkdir <- function(classifier){
  system(paste("mkdir data/data_",classifier[1],sep = ""))
}
getdir <- function(classifier){
  return(paste("data/data_",classifier[1],sep = ""))
}
isCNN <- function(classifier){
  return(as.logical(classifier[3]))
}
#Converts loaded data to PS format
convertFormat <- function(trainM){
  #sample.names <- trainM[,1] #Getnames of samples
  true.class <- as.numeric(trainM[,1])   #Get true class
  pred.class <- as.numeric(trainM[,2])   #Get predicted class
  pred.prob.raw <- trainM[,3:(3+number.classes-1)] #Get prediction probabilities
  pred.prob.numeric <- apply(pred.prob.raw,1,as.numeric) #Convert to numbers
  pred.prob <- t(apply(pred.prob.numeric, 2, function(x){return(x/sum(x))})) # Normalize probabilities
  #according to https://github.com/scikit-learn/scikit-learn/blob/0fb307bf3/sklearn/multiclass.py#L346
  return(list(
              #sample.names = sample.names, 
              true.label = true.class, 
              predicted.label = pred.class, 
              class.prob = pred.prob))
}
#Creates file to write
createFileData <- function(trainM){
  data.matrix.PS <- cbind(trainM$class.prob, trainM$predicted.label, trainM$true.label)
  colnames(data.matrix.PS) <- c(unlist(lapply(seq(0,number.classes-1),function(x){paste("probs",x,sep = "")})), "ptarg","ttarg")
  return(data.matrix.PS)
}
#Convert Relevance vector to PS format
convertARD <- function(RelVector){
  dim(RelVector) <- c(1,length(RelVector))
  colnames(RelVector) <- unlist(lapply(seq(1,length(RelVector)),function(x){return(paste("ard_",x,sep = ""))}))
  return(RelVector)
}
# The target format is: probs0, ..., probsn,ptarg,ttarg
number.classes <- 6 #Number og classes for GPC
k.fold <- seq(0,9)  #sequence of k's
iterations <- 10    #How many times repeated
isNoCNN <- F        #Flag indicates if classifier is CNN or not
#BGPLVM.replace.manual        <- c("BGPLVM_resampling_manual",       "../GPC/BGPLVM_replace/",       isNoCNN)
BGPLVM.noReplace.manual      <- c("BGPLVM_noResampling_manual",     "/home/mluser/git/NT_BodyParts/GPC/GPLVM/",     isNoCNN)
#BGPLVM.replace.top14         <- c("BGPLVM_resampling_top14",        "../GPC/BGPLVM_replace_14/",    isNoCNN)
BGPLVM.noReplace.top14       <- c("BGPLVM_noResampling_top14",      "/home/mluser/git/NT_BodyParts/GPC/GPLVM_full/",  isNoCNN)
#Procrustes.replace.manual    <- c("Procrustes_resampling",          "../GPC/Procrustes_replace/",   isNoCNN)
Procrustes.noReplace.manual  <- c("Procrustes_noResampling",        "/home/mluser/git/NT_BodyParts/GPC/Procrustes/", isNoCNN)
#CNN.replace.augmented        <- c("CNN_Aug_resampling",             "../CNN_augmentation/CNN_replace/",   !isNoCNN)
CNN.noReplace.augmented      <- c("CNN_Aug_noResampling",           "/home/mluser/git/NT_BodyParts/CNN/Classification/Augmented/", !isNoCNN)
#--- summarize classifier ---#
processor <- rbind(#BGPLVM.replace.manual,
                   BGPLVM.noReplace.manual,
                   #BGPLVM.replace.top14,
                   BGPLVM.noReplace.top14,
                   #Procrustes.replace.manual,
                   Procrustes.noReplace.manual,
                   #CNN.replace.augmented,
                   CNN.noReplace.augmented
                   )
system("rm -rf data/data*") #RM previous classifier results
#---------------------#
#--- Do processing ---#
#---------------------#
for(i in seq(1,nrow(processor))){ #Process all classifiers
  mkdir(processor[i,]) #Create directory for current classifier
  #---------------------------------#
  #--- Get classification result ---#
  #---------------------------------#
  REL.memory <- c()
  for(ITERATION in seq(0,iterations-1)){
    #if(!isCNN(processor[i,])){
    #  classifier.result <- suppressMessages(as.matrix(read_csv(paste(processor[i,2],ITERATION,"__dataMatrix.csv",sep = ""), col_names = FALSE)))
    #}else{
      classifier.result <- suppressMessages(as.matrix(read_csv(paste(processor[i,2],ITERATION,"_dataMatrix.csv",sep = ""))))
      #stop("Not implemented yet")
    #}
    #--- Get format ---#
    classifier.result.format <- convertFormat(classifier.result) #Convert to numeric and a list
    classifier.data_for_file <- createFileData(classifier.result.format) #Convert to PS format
    #--- Store ---#
    write.table(classifier.data_for_file, paste(getdir(processor[i,]),"/",ITERATION,"_classifierResult.csv",sep = ""),row.names = F,quote = F,sep = ',')
    #----------------------------------------------------------#
    #--- Get ls values and prepare for relevance processing ---#
    #----------------------------------------------------------#
    if(!isCNN(processor[i,])){
      REL.memory.local <- c() #Memory for LS values
      REL.memory.per_Population <- c()
      cat("Get GPC relevance for ", processor[i,1],"\n")
      for(k in k.fold){
        LS.k <- suppressMessages(as.matrix(read_table2(paste(processor[i,2],ITERATION,"_foldNr_",k,"_LS.csv",sep = ""), col_names = FALSE)))
        REL.k <- 1/LS.k
        REL.memory.local <- rbind(REL.memory.local,colMeans(REL.k))
        if(is.null(dim(REL.memory.per_Population))){
          REL.memory.per_Population <- REL.k
        }else{
          REL.memory.per_Population <- REL.memory.per_Population+REL.k
        }
      }
      REL.memory.per_Population <- REL.memory.per_Population / length(k.fold) #Normalize per-population relevance
      REL.memory <- rbind(REL.memory,colMeans(REL.memory.local)) # Just for later plotting
      write.table(convertARD(colMeans(REL.memory.local)), paste(getdir(processor[i,]),"/",ITERATION,"_Relevance.csv",sep = ""), row.names = F,quote = F,sep = ",")
      write.table(REL.memory.per_Population, paste(getdir(processor[i,]),"/",ITERATION,"_MeanRelevance_perPopulation.csv",sep = ""), row.names = F,quote = F,sep = ",")
    }#End if isCNN
  }#End Iteration loop
  #-------------------------------------#
  #--- Store results of REL analysis ---#
  #-------------------------------------#
  if(!isCNN(processor[i,])){
    pdf(paste(getdir(processor[i,]),"/Relevance.pdf",sep = ""))
    matplot(t(REL.memory),
            type='l', lty=1, col=seq(1,10), 
            xlab = "Features", ylab="Relevance",main=paste("Rel. Analysis for",processor[i,1]), 
            cex.main=0.75)
    abline(v = seq(0,40,1),h = seq(0,2,0.1),col='gray')
    dev.off()
    write.table(REL.memory, paste(getdir(processor[i,]),"/RelevanceSUM.csv",sep = ""),col.names = F, row.names = F,quote = F)
  }
}#End classifier loop